pkg/controller/association/service_account.go

```diff
@@ -0,0 +1,305 @@
+	serviceAccount commonv1.ServiceAccountName,
+	applicationName string,
+	applicationUID types.UID,
+) error {
+	tokenName := tokenName(applicationSecretName.Namespace, applicationName, applicationUID)
+	token, err := reconcileApplicationSecret(ctx, client, es, applicationSecretName, commonLabels, tokenName, serviceAccount)
+	if err != nil {
+		return err
+	}
+	return reconcileElasticsearchSecret(ctx, client, es, elasticsearchSecretName, commonLabels, *token)
+}
+
+// getCurrentApplicationToken returns the current token from the application Secret, or nil if the content of the Secret is not valid.
+func getCurrentApplicationToken(es *esv1.Elasticsearch, secretName string, secretData map[string][]byte) *Token {
+	if len(secretData) == 0 {
+		log.V(1).Info("secret is empty", "es_name", es.Name, "namespace", es.Namespace, "secret", secretName)
+		return nil
+	}
+	result := &Token{}
+	if value := getFieldOrNil(es, secretName, secretData, esuser.ServiceAccountTokenNameField); value != nil && len(*value) > 0 {
+		result.TokenName = *value
+	} else {
+		return nil
+	}
+
+	if value := getFieldOrNil(es, secretName, secretData, ServiceAccountTokenValueField); value != nil && len(*value) > 0 {
+		result.Token = *value
+	} else {
+		return nil
+	}
+
+	if value := getFieldOrNil(es, secretName, secretData, esuser.ServiceAccountHashField); value != nil && len(*value) > 0 {
+		result.Hash = *value
+	} else {
+		return nil
+	}
+
+	if value := getFieldOrNil(es, secretName, secretData, ServiceAccountNameField); value != nil && len(*value) > 0 {
+		result.ServiceAccountName = *value
+	} else {
+		return nil
+	}
+
+	return result
+}
+
+func getFieldOrNil(es *esv1.Elasticsearch, secretName string, secretData map[string][]byte, fieldName string) *string {
+	data, exists := secretData[fieldName]
+	if !exists {
+		log.V(1).Info(fmt.Sprintf("%s field is missing in service account token Secret", fieldName), "es_name", es.Name, "namespace", es.Namespace, "secret", secretName)
+		return nil
+	}
+	fieldValue := string(data)
+	return &fieldValue
+}
+
+var prefix = [...]byte{0x0, 0x1, 0x0, 0x1}
+
+// newApplicationToken generates a new token for a given service account.
+func newApplicationToken(serviceAccountName commonv1.ServiceAccountName, tokenName string) (*Token, error) {
+	secret := common.RandomBytes(64)
+	hash, err := pbkdf2Key(secret)
+	if err != nil {
+		return nil, err
+	}
+
+	fullyQualifiedName := fmt.Sprintf("%s/%s", Namespace, serviceAccountName)
+	suffix := []byte(fmt.Sprintf("%s/%s:%s", fullyQualifiedName, tokenName, secret))
+	token := base64.StdEncoding.EncodeToString(append(prefix[:], suffix...))
+
+	return &Token{
+		ServiceAccountName: fullyQualifiedName,
+		TokenName:          tokenName,
+		Token:              token,
+		Hash:               hash,
+	}, nil
+}
+
+func tokenName(
+	applicationNamespace, applicationName string,
+	applicationUID types.UID,
+) string {
+	return fmt.Sprintf("%s_%s_%s", applicationNamespace, applicationName, applicationUID)
```

Comment:
> Is it so that tokens get rotated when the application is recreated?

Yes, I want to be sure that different deployments get different credentials, even if they had the same name at some point.